### PR TITLE
Feature/sg 631 add registry collate funcs samplers

### PR DESCRIPTION
**Pull request by eran-deci**

### CODE REVIEW

File: src/super_gradients/common/registry/registry.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
 from super_gradients.training.models.segmentation_models.unet.unet_encoder import BACKBONE_STAGES
 from super_gradients.training.models.segmentation_models.unet.unet_decoder import UP_FUSE_BLOCKS
 from super_gradients.training.datasets.all_target_generators import ALL_TARGET_GENERATORS
+from super_gradients.training.datasets.all_collate_functions import ALL_COLLATE_FUNCTIONS
+from super_gradients.training.datasets.samplers.all_samplers import SAMPLERS
```
Comment (ofrimasad) [inline src/super_gradients/common/registry/registry.py:17]: please add the native torch samplers.
look for all classes that inherit from `Sampler` in `torch.utils.data.sampler`

**Review by ofrimasad — CHANGES_REQUESTED** [review]
minor comment

File: src/super_gradients/training/datasets/all_collate_functions.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from super_gradients.training.datasets.datasets_utils import ComposedCollateFunction, MultiScaleCollateFunction
+
+ALL_COLLATE_FUNCTIONS = {
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/all_collate_functions.py:3]: There are many more collate functions in SG. ctrl+alt+shift+f the project and add them please.

File: src/super_gradients/common/registry/registry.py [hunk 3]
```diff
@@ -59,3 +61,5 @@ def decorator(cls: Callable) -> Callable:
 register_unet_backbone_stage = create_register_decorator(registry=BACKBONE_STAGES)
 register_unet_up_block = create_register_decorator(registry=UP_FUSE_BLOCKS)
 register_target_generator = create_register_decorator(registry=ALL_TARGET_GENERATORS)
+register_collate_function = create_register_decorator(registry=ALL_COLLATE_FUNCTIONS)
```
Comment (shaydeci) [inline src/super_gradients/common/registry/registry.py:64]: Registering is essential but not sufficient...see my summary.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Comment regarding more collate functions left out, inline.

More importantly, the fact you added a registry for collate functions iis great but not sufficient.
Suppose a user registers their collate functions, and we pass some string as an argument to train_dataloader_params. Will that string translate to the right collate function?
The answer is no. You need to resolve the right class by implementing a factory otherwise the registry is useless.
Guidance: see dataloaders.get(), find where we handle the dataloader_params processing and resolve the collate_fn argument with your new CollateFactory.

File: src/super_gradients/common/registry/registry.py (hunk 1 above)
Comment (eran-deci) [reply src/super_gradients/common/registry/registry.py:17]: added in commit "updated_ fixes"

File: src/super_gradients/training/datasets/all_collate_functions.py (hunk 2 above)
Comment (eran-deci) [reply src/super_gradients/training/datasets/all_collate_functions.py:6]: added in commit "updated_ fixes"

File: src/super_gradients/common/registry/registry.py (hunk 3 above)
Comment (eran-deci) [reply src/super_gradients/common/registry/registry.py:64]: updated in commit "updated_ fixes"

File: src/super_gradients/training/datasets/samplers/all_samplers.py [hunk 4]
```diff
@@ -2,10 +2,16 @@
...
 from torch.utils.data.distributed import DistributedSampler
-
+from torch.utils.data.sampler import Sampler, SequentialSampler, SubsetRandomSampler, RandomSampler, WeightedRandomSampler, BatchSampler
 
 SAMPLERS = {
     Samplers.INFINITE: InfiniteSampler,
     Samplers.REPEAT_AUG: RepeatAugSampler,
-    Samplers.DISTRIBUTED: DistributedSampler
+    Samplers.DISTRIBUTED: DistributedSampler,
+    Samplers.SAMPLER: Sampler,
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/samplers/all_samplers.py:11]: This is a base class and will crash the code if trying to use it.

File: src/super_gradients/training/datasets/samplers/all_samplers.py [hunk 5]
```diff
@@ -2,10 +2,16 @@
...
+    Samplers.SEQUENTIAL: SequentialSampler,
+    Samplers.SUBSET_RANDOM: SubsetRandomSampler,
+    Samplers.RANDOM: RandomSampler,
+    Samplers.WEIGHTED_RANDOM: WeightedRandomSampler,
+    Samplers.BATCH: BatchSampler,
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/samplers/all_samplers.py:16]: Batch sampler is not a sampler...Its treated different and passed as a different argument to DataLoader constructor. See how we handle it inside datalaoders.get, and remove it from here please.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Almost there, just 2 more comments inline.

File: src/super_gradients/training/datasets/samplers/all_samplers.py (hunk 5 above)
Comment (eran-deci) [reply src/super_gradients/training/datasets/samplers/all_samplers.py]: removed

File: src/super_gradients/training/datasets/samplers/all_samplers.py (hunk 4 above)
Comment (eran-deci) [reply src/super_gradients/training/datasets/samplers/all_samplers.py]: removed

**Review by shaydeci — APPROVED** [review]
LGTM

**Review by ofrimasad — DISMISSED** [review]
LGTM

**Review by ofrimasad — APPROVED** [review]
first code contribution to SG - Good luck!

**Pull request merged**